### PR TITLE
BREAKING CHANGE: use MongoDB node driver 6, drop support for `rawResult` option and `findOneAndRemove()`

### PR DESCRIPTION
**Pull request by vkarpov15**
**Summary**

Use the latest commits to MongoDB Node driver's 6.0 branch to make sure Mongoose is compatible. 3 big changes:

1. No more `rawResult` option. This PR replaces `rawResult` with the `includeResultMetadata` option
2. Because of `includeResultMetadata`, we need to slightly restructure how `orFail()` works on `findOneAndX()`. Specifically, `orFail()` will now only error out when `findOneAndX()` doesn't return a document; previously, `orFail()` would also throw an error if `upsert` was set and a new document was upserted. With `includeResultMetadata`, detecting whether a document was upserted or not is no longer possible in general. However, I think this new approach makes more sense, or at least the TypeScript types make more sense, because TS types treat `orFail()` as making the query result non-nullable.
3. No more `findOneAndRemove()`. `findOneAndRemove()` was just a legacy wrapper around `findOneAndDelete()`, didn't even have its own middleware. 

**Examples**

### CODE REVIEW

**Comment by hasezoey** [comment]
> `git@github.com:mongodb/node-mongodb-native.git#6.0`

i think this PR should be marked as `Draft`(WIP), or is this intended to be merged like this into 8.0 until it comes out?

**Review by AbdelrahmanHafez — APPROVED** [review]
LGTM 👍

**Review by hasezoey — CHANGES_REQUESTED** [review]
mongodb 6.0.0 requires nodejs 16.20.1 or higher

**Comment by hasezoey** [comment]
also from [bson release notes](https://github.com/mongodb/js-bson/releases/tag/v6.0.0):
> Strings of length 12 can no longer make an ObjectId

tests:
>          strings of length 12 are valid oids (gh-3365):

>         custom casters:
>     CastError: Cast to ObjectId failed for value "12charstring" (type string) at path "myId" because of "BSONError"

>       with inheritance:
>Actual message: "Cast to ObjectId failed for value "12charstring" (type string) at path "test" because of "BSONError""

File: docs/migrating_to_8.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+you should be aware of when migrating from Mongoose 7.x to Mongoose 8.x.
+
+If you're still on Mongoose 6.x or earlier, please read the [Mongoose 6.x to 7.x migration guide](migrating_to_7.html) and upgrade to Mongoose 7.x first before upgrading to Mongoose 8.
+
+* [Removed `rawResult` option for `findOneAndUpdate()`](#removed-rawresult-option-for-findoneandupdate)
+* [Changed behavior for `findOneAndUpdate()` with `orFail()` and upsert](#changed-behavior-for-findoneandupdate-with-orfail-and-upsert)
+* [MongoDB Node Driver 6.0](#mongodb-node-driver-6)
+* [Removed `findOneAndRemove()`](#removed-findoneandremove)
+
+<h2 id="removed-rawresult-option-for-findoneandupdate"><a href="#removed-rawresult-option-for-findoneandupdate">Removed `rawResult` option for <code>findOneAndUpdate()</code></a></h2>
```
Comment (hasezoey) [inline docs/migrating_to_8.md:19]:
```suggestion
<h2 id="removed-rawresult-option-for-findoneandupdate"><a href="#removed-rawresult-option-for-findoneandupdate">Removed <code>rawResult</code> option for <code>findOneAndUpdate()</code></a></h2>
```

File: docs/migrating_to_8.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  new: true,
+  upsert: true,
+  // Replace `rawResult: true` with `includeResultMetadata: true`
+  includeResultMetadata: true
+});
+```
+
+`includeResultMetadata` in Mongoose 8 behaves identically to `rawResult`.
+
+<h2 id="mongodb-node-driver-6"><a href="#mongodb-node-driver-6">Changed behavior for <code>findOneAndUpdate()</code> with <code>orFail()</code> and upsert</a></h2>
```
Comment (hasezoey) [inline docs/migrating_to_8.md:37]:
```suggestion
<h2 id="changed-behavior-for-findoneandupdate-with-orfail-and-upsert"><a href="#changed-behavior-for-findoneandupdate-with-orfail-and-upsert">Changed behavior for <code>findOneAndUpdate()</code> with <code>orFail()</code> and upsert</a></h2>
```

**Review by hasezoey — CHANGES_REQUESTED** [review]
some minor markdown things

File: docs/migrating_to_8.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+Mongoose 8 uses [v6.x of the MongoDB Node driver](https://github.com/mongodb/node-mongodb-native/blob/main/HISTORY.md#600-2023-08-28).
+There's a few noteable changes in MongoDB Node driver v6 that affect Mongoose:
+
+1. The `ObjectId` constructor no longer accepts strings of length 12. In Mongoose 7, `new mongoose.Types.ObjectId('12charstring')` was perfectly valid. In Mongoose 8, `new mongoose.Types.ObjectId('12charstring')` throws an error.
+
+<h2 id="removed-findoneandremove"><a href="#removed-findoneandremove">Removed <code>findOneAndRemove()</code></a></h2>
+
+In Mongoose 7, `findOneAndRemove()` was an alias for `findOneAndDelete()` that Mongoose supported for backwards compatibility.
+Mongoose 8 no longer supports `findOneAndRemove()`.
+Use `findOneAndDelete()` instead.
```
Comment (hasezoey) [inline docs/migrating_to_8.md:56]:
```suggestion
Use `findOneAndDelete()` instead.

```

markdownlint

Comment (vkarpov15) [reply docs/migrating_to_8.md]: Fixed in 9eccbb6

File: docs/migrating_to_8.md (hunk 2 above)
Comment (vkarpov15) [reply docs/migrating_to_8.md]: Fixed in 9eccbb6

**Comment by vkarpov15** [comment]
@hasezoey fixed, can you please re-review?

File: test/model.test.js [hunk 4]
```diff
@@ -5050,10 +5050,6 @@ describe('Model', function() {
       await Model.createCollection();
       const collectionName = Model.collection.name;
 
-      // If the collection is not created, the following will throw
-      // MongoServerError: Collection [mongoose_test.User] not found.
-      await db.collection(collectionName).stats();
-
```
Comment (hasezoey) [inline test/model.test.js:5056]: what is this change? is it not necessary anymore?

**Review by hasezoey — APPROVED** [review]
looks good to me now.

i dont know if it is related to this PR, but some replica set tests are timing out, and one test somehow tried to run while drop is occuring

File: test/model.test.js (hunk 4 above)
Comment (vkarpov15) [reply test/model.test.js:5056]: I don't think that line is necessary for the test. And `stats()` is no longer a method on collections.

**Comment by vkarpov15** [comment]
@hasezoey I took a closer look and the test failures were indicative of a bug where we weren't passing `session` option to `Document.prototype.deleteOne` correctly, that's fixed now :+1:

**Pull request merged**